src/System.IO.Pipelines/Pipe.cs

```diff
@@ -509,13 +509,17 @@ bool IPipeReader.TryRead(out ReadResult result)
                 }
 
                 result = new ReadResult();
-                if (!_readerAwaitable.IsCompleted)
+                if (_length > 0 || _readerAwaitable.IsCompleted)
                 {
-                    return false;
+                    GetResult(ref result);
+                    return true;
                 }
 
-                GetResult(ref result);
-                return true;
+                if(_readerAwaitable.WasAwaited)
```

Comment:
changed to Changed to HasContinuation. IsAwaiting isn't logical as it might not be awaiting anymore if the awaiter is completed and this property doesn't take that into account